File: osx/BubblemonView.m
//
//  BubblemonView.m
//  bubblemon
//
//  Created by Johan Walles on 2012-04-29.
//  Copyright (c) 2012 __MyCompanyName__. All rights reserved.
//

#import "BubblemonView.h"

@implementation BubblemonView

- (void)setDockTile:(NSDockTile*)aDockTile {
  if (aDockTile) {
    dockTile = aDockTile;
    [dockTile setContentView:self];
  }
}

- (NSMenu *)dockMenu {
  NSLog(@"Setting up dock menu\n");
  if (dockMenu != NULL) {
    return dockMenu;
  }
  
  dockMenu = [[NSMenu alloc] init];
  
  NSMenuItem *helpItem = [[NSMenuItem alloc] initWithTitle:@"Help" action:@selector(openLegend:) keyEquivalent:@""];
  [helpItem setTarget:self];
  [dockMenu addItem:helpItem];
  
  NSMenuItem *aboutItem = [[NSMenuItem alloc] initWithTitle:@"About" action:@selector(openAboutPanel:) keyEquivalent:@""];
  [aboutItem setTarget:self];
  [dockMenu addItem:aboutItem];
  
  return dockMenu;
}

- (IBAction)openLegend:(id)sender {
  NSLog(@"Opening help in browser...\n");
  [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:@"http://walles.github.io/bubblemon/"]];
}

- (IBAction)openAboutPanel:(id)sender {
  NSAttributedString *credits = [[NSAttributedString alloc]
                                 initWithString:@"http://walles.github.io/bubblemon"
                                 attributes: @{NSLinkAttributeName: @"http://walles.github.io/bubblemon"}];
  
  // The Git hash and version get filled in by a "Run Script" build step
  NSBundle *bundle = [NSBundle bundleForClass:[self class]];
  NSString *gitHash = [bundle infoDictionary][@"GitHash"];
  NSString *gitDescribe = [bundle infoDictionary][@"GitDescribe"];
  
  NSDictionary *aboutOptions = @{
                                 @"Credits": credits,
                                 @"ApplicationName": @"Bubblemon",
                                 // @"ApplicationIcon": FIXME,
                                 @"Version": gitHash,
                                 @"Copyright": @"Copyright 1999-2014 johan.walles@gmail.com",
                                 @"ApplicationVersion": gitDescribe
                                 };
  [[NSApplication sharedApplication] orderFrontStandardAboutPanelWithOptions: aboutOptions];
}

- (id)initWithFrame:(NSRect)frame
{
  self = [super initWithFrame:frame];
  if (self) {
    // Initialization code here.
#ifdef DEBUG
    bubblemon_selftest();
#endif
    bubblemon = bubblemon_init();
    bubblemon_setColors(bubblemon,
                        0x75ceff00u, 0x0066ff80u,
                        0xff333340u, 0xaa000080u,
                        0x00ff0080u, 0xffff40ffu);
    picture = NULL;
    
    // Inspired by http://stackoverflow.com/questions/1449035/how-do-i-use-nstimer
    [NSTimer scheduledTimerWithTimeInterval:(1.0 / 25.0)
                                     target:self
                                   selector:@selector(timerFired)
                                   userInfo:nil
                                    repeats:YES];
    
    // Load window frame graphics
    NSBundle *bundle = [NSBundle bundleForClass:[self class]];
    CFURLRef windowFrameUrl = (__bridge CFURLRef)[bundle URLForImageResource:@"window-frame"];
    CGDataProviderRef dataProvider = CGDataProviderCreateWithURL(windowFrameUrl);
    windowFrame = CGImageCreateWithPNGDataProvider(dataProvider, NULL, NO, kCGRenderingIntentDefault);
    CGDataProviderRelease(dataProvider);
  }
  
  return self;
}

- (void)timerFired
{
  // Compute a new image to display
  // The Dock won't tell us its size, so this is a guess at roughly how many pixels
  // the bubblemon will get on screen.
  bubblemon_setSize(bubblemon, 50, 50);
  picture = bubblemon_getPicture(bubblemon);
  
  NSString *tooltip = [[NSString alloc] initWithUTF8String:bubblemon_getTooltip(bubblemon)];
  [self setToolTip: tooltip];
  
  if (dockTile != nil) {
    [dockTile display];
  } else {
    [self setNeedsDisplay:YES];
  }
}

static void releaseDataProvider(void *info, const void *data, size_t size) {
  // No need to actually free anything here, the data is handled by
  // the bubblemon code.
}

- (void)drawRect:(NSRect)dirtyRectIgnored
{
  if (!picture) {
    return;
  }
  
  // Redraw ourselves
  size_t imageDataSize = picture->width * picture->height * sizeof(bubblemon_color_t);
  
  CGDataProviderRef dataProviderRef = CGDataProviderCreateWithData(NULL, picture->pixels, imageDataSize, releaseDataProvider);
  
  const size_t bitsPerComponent = 8;
  const size_t bitsPerPixel = sizeof(bubblemon_color_t) * 8;
  const size_t bytesPerRow = picture->width * sizeof(bubblemon_color_t);
  const bool shouldInterpolate = false;
  
  CGColorSpaceRef rgb = CGColorSpaceCreateDeviceRGB();
  CGImageRef cgImageRef = CGImageCreate(picture->width,
                                        picture->height,
                                        bitsPerComponent,
                                        bitsPerPixel,
                                        bytesPerRow,
                                        rgb,
                                        (CGBitmapInfo)kCGImageAlphaNoneSkipLast,
                                        dataProviderRef,
                                        NULL,
                                        shouldInterpolate,
                                        kCGRenderingIntentDefault);
  CGColorSpaceRelease(rgb);
  CGDataProviderRelease(dataProviderRef);
  
  __strong NSGraphicsContext *nsGraphicsContext = [NSGraphicsContext currentContext];
  CGContextRef cgContextRef = (CGContextRef)[nsGraphicsContext graphicsPort];
  
  // Draw the bubblemon image
  CGRect bubbleViewRect = CGRectMake([self bounds].size.width  * 0.08f,
                                     [self bounds].size.height * 0.08f,
                                     [self bounds].size.width  * 0.84f,
                                     [self bounds].size.height * 0.84f);
  CGContextSetAlpha(cgContextRef, 0.9f);
  CGContextDrawImage(cgContextRef, bubbleViewRect, cgImageRef);
  CGImageRelease(cgImageRef);
  
  // Draw the window frame
  CGRect fullSizeRect = NSRectToCGRect([self bounds]);
  CGContextSetAlpha(cgContextRef, 1.0f);
  CGContextDrawImage(cgContextRef, fullSizeRect, windowFrame);
}

@end

File: osx/BubblemonView.swift
import Cocoa

class BubblemonView: NSView, NSDockTilePlugIn {
  private var _bubblemon: UnsafeMutablePointer<bubblemon_t>
  private var _picture: UnsafePointer<bubblemon_picture_t>?
  private var _dockTile: NSDockTile?
  private var _dockMenu: NSMenu?
  private var _windowFrame: CGImage?
  private var _scaledWindowFrame: CGLayer?

  private var _touchBarMode: Bool

  func setDockTile(_ dockTile: NSDockTile?) {
    if dockTile != nil {
      _dockTile = dockTile
      _dockTile!.contentView = self
    }
  }

  func dockMenu() -> NSMenu? {
    NSLog("Setting up dock menu")

    if _dockMenu != nil {
      return _dockMenu
    }

    let menu = NSMenu()
    let helpItem = NSMenuItem(title: "Help", action: #selector(self.openLegend), keyEquivalent: "")
    helpItem.target = self
    menu.addItem(helpItem)
    let aboutItem = NSMenuItem(title: "About", action: #selector(self.openAboutPanel), keyEquivalent: "")
    aboutItem.target = self
    menu.addItem(aboutItem)
    let shareItem = NSMenuItem(title: "Share on Facebook", action: #selector(self.shareOnFacebook), keyEquivalent: "")
    shareItem.target = self
    menu.addItem(shareItem)

    // From: https://stackoverflow.com/a/55403087/473672
    let submenu = NSMenu()
    let mainDropdown = NSMenuItem(title: "Debug", action: nil, keyEquivalent: "")
    menu.addItem(mainDropdown)
    menu.setSubmenu(submenu, for: mainDropdown)

    let showNumbersItem = NSMenuItem(title: "Show Numbers", action:
      #selector(self.showLoadNumbers), keyEquivalent: "")
    showNumbersItem.target = self
    submenu.addItem(showNumbersItem)

    _dockMenu = menu
    return _dockMenu
  }

  @IBAction func openLegend(_ sender: Any) {
    NSLog("Opening help in browser...")
    // From: http://lists.apple.com/archives/xcode-users/2016/Feb/msg00111.html
    let url = URL(string: "https://walles.github.io/bubblemon/")
    NSWorkspace.shared.open(url!)
  }

  @IBAction func shareOnFacebook(_ sender: Any) {
    NSLog("Opening browser to share on Facebook...")
    // From: http://lists.apple.com/archives/xcode-users/2016/Feb/msg00111.html
    let url = URL(string: "https://www.facebook.com/sharer/sharer.php?u=http%3A//walles.github.io/bubblemon/")
    NSWorkspace.shared.open(url!)
  }

  @IBAction func openAboutPanel(_ sender: Any) {
    let credits = NSAttributedString(
      string: "https://walles.github.io/bubblemon",
      attributes: [
        NSAttributedString.Key.link: "https://walles.github.io/bubblemon"
      ]
    )

    // The Git hash and version get filled in by a "Run Script" build step
    let bundle = Bundle(for: BubblemonView.self)
    let version = bundle.infoDictionary?["CFBundleShortVersionString"] as? String
    let gitHash = bundle.infoDictionary?["GitHash"] as? String
    let icon = bundle.image(forResource: "icon.png.icns")
    let aboutOptions: [NSApplication.AboutPanelOptionKey: Any] = [
      NSApplication.AboutPanelOptionKey.applicationName: "Bubblemon",
      NSApplication.AboutPanelOptionKey.applicationVersion: version!,
      NSApplication.AboutPanelOptionKey.version: gitHash!,
      NSApplication.AboutPanelOptionKey.applicationIcon: icon!,
      NSApplication.AboutPanelOptionKey.credits: credits,
    ]
    NSApplication.shared.activate(ignoringOtherApps: true)
    NSApplication.shared.orderFrontStandardAboutPanel(options: aboutOptions)
  }

  @IBAction func showLoadNumbers(_ sender: Any) {
    let alert = NSAlert()

    alert.messageText = String(cString: bubblemon_getTooltip(_bubblemon)!)

    alert.alertStyle = NSAlert.Style.informational
    alert.addButton(withTitle: "OK")

    // This might freeze the physics, but give us the numbers. Call it an MVP!
    alert.runModal()
  }

  override init(frame: NSRect) {
    _touchBarMode = false

#if DEBUG
    bubblemon_selftest()
#endif
    _bubblemon = bubblemon_init()
    _picture = nil
    // Load window frame graphics
    let bundle = Bundle(for: BubblemonView.self)
    let windowFrameUrl = bundle.urlForImageResource("window-frame")
    let windowFrameNsUrl = NSURL(string: (windowFrameUrl?.absoluteString)!)

    let dataProvider = CGDataProvider(url: windowFrameNsUrl!)
    _windowFrame = CGImage(
      pngDataProviderSource: dataProvider!,
      decode: nil,
      shouldInterpolate: false,
      intent: CGColorRenderingIntent.defaultIntent)

    super.init(frame: frame)

    // Inspired by http://stackoverflow.com/questions/1449035/how-do-i-use-nstimer
    Timer.scheduledTimer(
      timeInterval: (1.0 / 10.0),
      target: self,
      selector: #selector(self.timerTriggered),
      userInfo: nil,
      repeats: true)
  }

  required init?(coder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }

  func setTouchBarMode(_ touchBarMode: Bool) {
    _touchBarMode = touchBarMode
  }

  func getCachedWindowFrame(baseContext: CGContext) -> CGLayer {
    // Can we use our cached window frame?
    if _scaledWindowFrame == nil {
      // No; we don't have one
    } else if (!bounds.size.equalTo(_scaledWindowFrame!.size)) {
      // No; the size doesn't match
    } else {
      // Yes!
      return _scaledWindowFrame!
    }

    // All of these numbers are because we want to keep the frame width the same
    // all around the frame, even if the frame is wider than it is high.
    //
    // The way we do this is by scaling the width in three parts. The leftmost
    // third is scaled the same in both directions, as is the rightmost third.
    //
    // The middle third of the original image is then scaled to cover the gap in
    // between the right and left parts. x1 and x2 are one and two thirds into
    // the frame image respectively.
    let yFactor = bounds.size.height / CGFloat(_windowFrame!.height)
    let x1before = CGFloat(_windowFrame!.width / 3)
    let x2before = CGFloat((2 * _windowFrame!.width) / 3)
    let middleWidthBefore = x2before - x1before
    let rightWidthBefore = CGFloat(_windowFrame!.width - 1) - x2before
    let heightBefore = CGFloat(_windowFrame!.height)

    let x1after = x1before * yFactor
    let x2after = bounds.size.width - 1 - x1after
    let middleWidthAfter = 1 + x2after - x1after
    let rightWidthAfter = bounds.size.width - 1 - x2after
    let heightAfter = bounds.size.height

    let scaledLayer = CGLayer(baseContext, size: bounds.size, auxiliaryInfo: nil)
    let scaledContext = scaledLayer!.context

    // Draw the left part
    scaledContext!.draw(
      _windowFrame!.cropping(to: CGRect(x: CGFloat(0), y: CGFloat(0), width: x1before, height: heightBefore))!,
      in: CGRect(x: 0, y: 0, width: x1after + 1, height: heightAfter))

    // Draw the middle part
    scaledContext!.draw(
      _windowFrame!.cropping(to: CGRect(
        x: x1before, y: CGFloat(0), width: middleWidthBefore, height: heightBefore))!,
      in: CGRect(x: x1after, y: CGFloat(0), width: middleWidthAfter, height: heightAfter))

    // Draw the right part
    scaledContext!.draw(
      _windowFrame!.cropping(to: CGRect(
        x: x2before, y: 0, width: rightWidthBefore, height: heightBefore))!,
      in: CGRect(x: x2after, y: CGFloat(0), width: rightWidthAfter, height: heightAfter))

    _scaledWindowFrame = scaledLayer!
    return _scaledWindowFrame!
  }

  @objc
  func timerTriggered() {
    if (!self.window!.occlusionState.contains(.visible)) {
      // Ref: https://developer.apple.com/forums/thread/71171
      return
    }

    // Compute a new image to display
    if (_touchBarMode) {
      // Touch Bar is tiny, make few bubbles and scale everything up to be big
      bubblemon_setSize(_bubblemon, 60, 30)
    } else {
      // The Dock won't tell us its size, so this is a guess at roughly how many pixels
      // the bubblemon will get on screen.
      bubblemon_setSize(_bubblemon, 50, 50)
    }
    _picture = bubblemon_getPicture(_bubblemon)
    let tooltip = String(utf8String: bubblemon_getTooltip(_bubblemon))
    toolTip = tooltip
    if _dockTile != nil {
      _dockTile!.display()
    } else {
      needsDisplay = true
    }
  }

  override func setFrameSize(_ newSize: NSSize) {
    if(_touchBarMode) {
      // Landscape mode on the touchbar because it is tiny
      let width = min(newSize.height * 2, newSize.width)
      let height = newSize.height

      // Center ourselves horizontally
      setFrameOrigin(NSPoint.init(x: (newSize.width - width) / 2, y: 0))

      super.setFrameSize(NSSize.init(width: width, height: height))
    } else {
      super.setFrameSize(newSize)
    }
  }

  override func draw(_ dirtyRect: NSRect) {
    if _picture == nil {
      return
    }
    let picture = _picture!.pointee

    // Redraw ourselves
    let imageDataSize = Int(picture.width * picture.height) * MemoryLayout<bubblemon_color_t>.size
    let dataProviderRef = CGDataProvider(
      dataInfo: nil,
      data: picture.pixels,
      size: imageDataSize,
      releaseData: releaseDataProvider)

    let bitsPerComponent: size_t = 8
    let bitsPerPixel = MemoryLayout<bubblemon_color_t>.size * 8
    let bytesPerRow = Int(picture.width) * MemoryLayout<bubblemon_color_t>.size
    let shouldInterpolate: Bool = false
    let rgb = CGColorSpaceCreateDeviceRGB()
    let cgImage = CGImage(
      width: Int(picture.width), height: Int(picture.height),
      bitsPerComponent: bitsPerComponent,
      bitsPerPixel: bitsPerPixel,
      bytesPerRow: bytesPerRow,
      space: rgb,
      bitmapInfo: CGBitmapInfo(rawValue: CGImageAlphaInfo.noneSkipLast.rawValue),
      provider: dataProviderRef!,
      decode: nil,
      shouldInterpolate: shouldInterpolate,
      intent: CGColorRenderingIntent.defaultIntent)

    let nsGraphicsContext = NSGraphicsContext.current
    guard let cgContext = nsGraphicsContext?.cgContext else {
      fatalError("unable to get a CGContext, can't draw")
    }

    // Draw the bubblemon image

    // The frame width is the same on all sides, and determined by the height of
    // the image.
    let frameWidth = 0.08 * bounds.size.height

    let bubbleViewRect = CGRect(
      x: frameWidth,
      y: frameWidth,
      width: bounds.size.width - 2.0 * frameWidth,
      height: bounds.size.height - 2.0 * frameWidth)
    cgContext.setAlpha(1.0)
    cgContext.interpolationQuality = .none
    cgContext.draw(cgImage!, in: bubbleViewRect)

    if (!_touchBarMode) {
      // Draw the window frame
      cgContext.setAlpha(1.0)
      cgContext.draw(getCachedWindowFrame(baseContext: cgContext), at: CGPoint(x: 0, y: 0))
    }
  }
}

func releaseDataProvider(_ context:UnsafeMutableRawPointer?, data:UnsafeRawPointer, size:Int) {
  // No need to actually free anything here, the data is handled by
  // the bubblemon code.
}
